python/lsst/pipe/base/_status.py

```diff
@@ -86,6 +93,123 @@ class RepeatableQuantumError(RuntimeError):
+    """
+
+
+class AnnotatedPartialOutputsError(RepeatableQuantumError):
+    """Exception that runQuantum raises when the (partial) outputs it has
+    written contain information about their own incompleteness or degraded
+    quality.
+
+    This exception should always chain the original error. When the
+    executor catches this exception, it will report the original exception. In
+    contrast, other exceptions raised from ``runQuantum`` are considered to
+    invalidate any outputs that are already written.
+    """
+
+    @classmethod
+    def annotate(
+        cls, error: Exception, *args: GetSetDictMetadataHolder | None, log: logging.Logger
+    ) -> AnnotatedPartialOutputsError:
+        """Set metadata on outputs to explain the nature of the failure.
+
+        Parameters
+        ----------
+        error : `Exception`
+            Exception that caused the task to fail.
+        *args : `GetSetDictMetadataHolder`
+            Objects (e.g. Task, Exposure, SimpleCatalog) to annotate with
+            failure information. They must have a `metadata` property.
+        log : `logging.Logger`
+            Log to send error message to.
+
+        Returns
+        -------
+        error : `AnnotatedPartialOutputsError`
+            Exception that the failing task can ``raise from`` with the
+            passed-in exception.
+
+        Notes
+        -----
+        This should be called from within an except block that has caught an
+        exception. Here is an example of handling a failure in
+        ``PipelineTask.runQuantum`` that annotates and writes partial outputs:
+
+        .. code-block:: py
+            :name: annotate-error-example
+
+            def runQuantum(self, butlerQC, inputRefs, outputRefs):
+                inputs = butlerQC.get(inputRefs)
+                exposures = inputs.pop("exposures")
+                assert not inputs, "runQuantum got more inputs than expected"
+
+                result = pipeBase.Struct(catalog=None)
+                try:
+                    self.run(exposure)
+                except pipeBase.AlgorithmError as e:
+                    error = pipeBase.AnnotatedPartialOutputsError.annotate(
+                        e, self, result.catalog, log=self.log
+                    )
+                    raise error from e
+                finally:
+                    butlerQC.put(result, outputRefs)
+        """
+        failure_info = {
+            "message": str(error),
+            "type": introspection.get_full_type_name(error),
+        }
+        if other := getattr(error, "metadata", None):
+            failure_info["metadata"] = other
+
+        # NOTE: Can't fully test this in pipe_base because afw is not a
+        # dependency; test_calibrateImage.py in pipe_tasks gives more coverage.
+        for item in args:
+            # Some outputs may not exist, so we cannot set metadata on them.
+            if item is None:
```

Comment:
Same Q as above, can this still have the desired behavior with the syntax `if not item` ?